In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
import re 

import plotly.plotly as py
import plotly.graph_objs as go
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

from IPython.display import Image
from IPython.core.display import HTML 

import warnings
warnings.filterwarnings("ignore")

init_notebook_mode(connected=True)

I am using shark attack incidents dataset from Kaggle: 
https://www.kaggle.com/teajay/global-shark-attacks. This data was compiled by the Global Shark Attack File http://www.sharkattackfile.net/.

Data was cleaned and preprocesseed in a separate Jupyter Notebook

In [16]:
sharks = pd.read_csv('shark_attack_cleaned.csv')

In [17]:
sharks.head(2)

Date  Year        Type Country        Area  \
0  25-Jun-2018  2018        Boat     USA  California   
1  18-Jun-2018  2018  Unprovoked     USA     Georgia   

                         Location  Activity Sex Age  \
0     Oceanside, San Diego County  Paddling   F  57   
1  St. Simon Island, Glynn County  Standing   F  11   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   

            Time      Species  Hour  Month  Hemisphere  
0          18h00  White shark  18.0    5.0         0.0  
1  14h00  -15h00          NaN  14.0    5.0         0.0

I'm going to use this color pallette for my plots:

In [9]:
light_blue = 'rgb(142, 212, 229)'
dark_blue = 'rgb(19, 77, 102)'
green = 'rgb(199, 204, 118)'
pink = 'rgb(254, 207, 173)'
orange = 'rgb(253, 174, 97)'
red = 'rgb(253, 107, 97)'

# How dangerous are sharks and is all the fear people have justified?

Most people are afraid of sharks, and there's a misconception that if you are in the water with them you are in deep trouble. The idea that sharks are killing machines preying on humans is a myth that I learned to overcome after I became an avid scuba diver, but I've only had two encounters with sharks so far. I’m often diving in the murky waters of Northern California, which is famous for its large population of white sharks, so I wanted to learn more about the attacks and real risks.

## Shark Attacks worldwide 1900-2017

Let’s take a look at the general development of shark attacks from 1900 to 2017.

In [22]:
fatal_attack = sharks[sharks['Fatal (Y/N)'] == 'Y'].groupby(['Year']).count().iloc[:, :1]
fatal_attack.rename(index=str, columns={"Date": "Fatal"}, inplace=True)

nonfatal_attack = sharks[sharks['Fatal (Y/N)'] == 'N'].groupby(['Year']).count().iloc[:, :1]
nonfatal_attack.rename(index=str, columns={"Date": "Non-Fatal"}, inplace=True)

fatality_unknown = sharks[sharks['Fatal (Y/N)'] == 'UNKNOWN'].groupby(['Year']).count().iloc[:, :1]
fatality_unknown.rename(index=str, columns={"Date": "Unknown"}, inplace=True)

scatter = pd.concat([fatal_attack,nonfatal_attack, fatality_unknown], axis=1).iloc[:-1, :]

In [23]:
scatter.tail(2)

Fatal  Non-Fatal  Unknown
2016    6.0        113     12.0
2017    8.0        115     14.0

In [24]:
scatter.Fatal.max()

24.0

In [25]:
fatal = go.Scatter(
    x = scatter.index,
    y = scatter['Fatal'],
    name = 'Fatal',
    mode = 'markers',
    marker = dict(
        size = 10,
        line = dict(
            width = 1,
            color = 'rgb(0, 0, 0)'
        ), color = pink))

nonfatal = go.Scatter(
    x = scatter.index,
    y = scatter['Non-Fatal'],
    name = 'Non-fatal',
    mode = 'markers',
    marker = dict(
        size = 10,
        line = dict(
            width = 1), color = green))
fatality_na = go.Scatter(
    x = scatter.index,
    y = scatter['Unknown'],
    name = 'Unknown',
    mode = 'markers',
    marker = dict(
        size = 10,
        line = dict(
            width = 1), color = light_blue))

data = [nonfatal, fatal, fatality_na]

layout = dict(title = 'Shark attacks worldwide 1900-2017',
              yaxis = dict(zeroline = False, title='Count of total shark attachs'),
              xaxis = dict(zeroline = False, title='Year'))

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Shark attacks scatter plot")

#plotly.offline.plot(data, filename='test.html') # creates html
# plotly.offline.plot(data, include_plotlyjs=False, output_type='div') # for jinja2 embeding

There is a peak in the attacks in around 1960, let's see where the peack is coming from

In [14]:
sharks[sharks['Year'] == 1960].groupby('Country').size().sort_values(ascending=False)[:5]

Country
USA                 25
AUSTRALIA           21
PAPUA NEW GUINEA    10
SOUTH AFRICA         9
IRAQ                 3
dtype: int64

In [15]:
sharks[sharks['Year'] == 1958].groupby('Country').size().sort_values(ascending=False)[:5]

Country
USA                 16
PAPUA NEW GUINEA     9
SOUTH AFRICA         8
AUSTRALIA            5
BAHAMAS              3
dtype: int64

Generally, shark attacks are increasing, though the fatal attacks are staying at the same level throughout the century, at about average of 9.2 number of fatal attacks globally per year. This number doesn’t change much by decade either.
<br>If you look at the shark attacks development, especially non-fatal ones, you can see a peak around 1960. The peak happens in the USA and Australia, driven by surfing starting to get popular in the 60s.

<i>Disclaimer: all following plots use the data from 1900 to 2018 unless otherwise commented.</i>

## Shark attacks by time of the day

Sharks hunt at dawn and dusk to benefit from darkness by using not only their highly evolved sense of smell but also detecting electricity and vibrations in the water. If sharks prey on humans, we should see a peak of attacks around dusk and dawn.

In [26]:
time_nonfatal = sharks[sharks['Fatal (Y/N)'] == 'N']
time_fatal = sharks[sharks['Fatal (Y/N)'] == 'Y']
time_na_fatality = sharks[sharks['Fatal (Y/N)'] == 'UNKNOWN']

In [27]:
hour = list(range(0,25,1))
hourtext = ['midnight', '1 am', '2 am', '3 am', '4 am', '5 am', '6 am', '7 am', '8 am', '9 am', '10 am', '11 am',
           'noon', '1 pm', '2 pm', '3 pm', '4 pm', '5 pm', '6 pm', '7 pm', '8 pm', '9 pm', '10 pm', '11 pm']
trace1 = go.Histogram(
    x=time_nonfatal['Hour'],
    opacity=1, name = "Non-fatal", marker=dict(color=green)
)
trace2 = go.Histogram(
    x=time_fatal['Hour'],
    opacity=1, name = "Fatal", marker=dict(color=pink)
)

trace3 = go.Histogram(
    x=time_na_fatality['Hour'],
    opacity=1, name = "Unknown", marker=dict(color=light_blue)
)

data = [trace1, trace2, trace3]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Time of the day',
            tickvals=list(range(0,25,1)),
            ticktext = hourtext, tickangle=-45),
                   
    title='Number of shark attacks by hour',
    yaxis=dict(
        title='Count of total shark attacks'
    ),
    bargap=0.1
)
fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks histogram")

Not true! Obviously, humans are not part of sharks’ diet and they are not targetting human, so the attacks happen when more people are in the water, between around 10 am and 5 pm.

## Number of shark attacks by type

The myth of the shark attack hunting people is scary. Let’s look at the data to find out how sharks attack.
From this plot, you can see that most of the time, the attack is unprovoked. Though, I believe the attacks happen when a shark mistakes people for their prey, as we will see in later charts showing the number of attacks per activity. For example, a surfer sitting on their surfboard waiting for a wave looks and behave like a hurt Seal or Sea Lion from the bottom.

In [28]:
sharks.Type.unique()

array(['Boat', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster'], dtype=object)

In [29]:
fatal_attack = sharks[sharks['Fatal (Y/N)'] == 'Y'].groupby(['Type']).count().iloc[:, :1]
fatal_attack.rename(index=str, columns={"Date": "Fatal"}, inplace=True)

nonfatal_attack = sharks[sharks['Fatal (Y/N)'] == 'N'].groupby(['Type']).count().iloc[:, :1]
nonfatal_attack.rename(index=str, columns={"Date": "Non-Fatal"}, inplace=True)

fatality_unknown = sharks[sharks['Fatal (Y/N)'] == 'UNKNOWN'].groupby(['Type']).count().iloc[:, :1]
fatality_unknown.rename(index=str, columns={"Date": "Unknown"}, inplace=True)

In [30]:
attacks_type = pd.concat([fatal_attack,nonfatal_attack, fatality_unknown], axis=1).sort_values('Non-Fatal', ascending=True)
attacks_type.index

Index(['Questionable', 'Invalid', 'Sea Disaster', 'Boat', 'Provoked',
       'Unprovoked'],
      dtype='object', name='Type')

In [31]:
trace1 = go.Bar(
    y=attacks_type.index,
    x=attacks_type['Unknown'],
    name='Unknown',
    orientation = 'h', marker=dict(color = light_blue))

trace2 = go.Bar(
    y=attacks_type.index,
    x=attacks_type['Non-Fatal'],
    name='Non-Fatal',
    orientation = 'h',
    marker = dict(color = green))

trace3 = go.Bar(
    y=attacks_type.index,
    x=attacks_type['Fatal'],
    name='Fatal',
    orientation = 'h',
    marker = dict(color = pink))

data = [trace1, trace2, trace3]

layout = go.Layout(barmode='stack',
        xaxis=dict(title='Number of attacks'),
                   
    title='Number of shark attacks by type',
    yaxis=dict(
        title='Type'
    )
)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks barplot")

We can see that about ⅓ of all attacks are either provoked or happening while people are on the Boat or at Sea Disaster or when the type is Invalid/Unknown. The boat attacks in the data set are usually when the shark bumps the boat. Sea Disaster is usually suspected that the shark was feasting on cadaver after people were drawn. This shows us that only in ⅔ of the cases there is an unprovoked attack. 

## Age distribution of shark attack victims

Let’s take a look at the age distribution of shark attack victims.

In [32]:
male = sharks[sharks.Sex=='M'].Age.values
female = sharks[sharks.Sex=='F'].Age.values

In [33]:
trace0 = go.Box(x=male, name='Male', marker = dict(
        color = dark_blue
    ))
trace1 = go.Box(x=female, name='Female', marker = dict(
        color = pink
    ))
data = [trace0, trace1]

layout = go.Layout(title = 'Age distribution of victim at shark attack',
    yaxis=dict(
        title='Gender',
        zeroline=False
    ), xaxis=dict(title='Age')
)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks barplot")

Seems that female victim age distribution is wider than the male one and the medium age is a bit lower. Overall, I think this chart reflects the age distribution of people who are active in water sports earlier in their life.

## Number of shark attacks by activity and country

We have seen a peak in the shark attacks during the ’60s and I assumed that it strongly correlates with the increasing popularity of surfing. Let’s analyze if it is true and what other activities people are engaged in when they are attacked by a shark. I also wanted to see a breakdown by country, thus I analyzed the top 15 countries by the count of shark attacks.

In [34]:
top_activities = list(sharks.groupby('Activity').count().sort_values('Date', ascending=False)[:10].index)

In [35]:
top_countries = list(sharks.groupby('Country').count().sort_values('Date', ascending=False)[:15].index)

In [36]:
heatmap = sharks[sharks.Activity.isin(top_activities) & 
           sharks.Country.isin(top_countries)]
heatmap['Count'] =1
heatmap = heatmap[['Country', 'Activity', 'Count']]
a = heatmap.groupby(['Country', 'Activity'], group_keys=False).sum()
heatmap_dict = {}
for i in a.itertuples():
    heatmap_dict[i[0]] =i[1]

In [37]:
heatmap_dict[('USA', 'Surfing')]

655

In [38]:
z = []

for act in top_activities:
    new_row = []
    for country in top_countries:
        if (str(country), str(act)) in heatmap_dict:
            new_row.append(heatmap_dict[(str(country), str(act))])
        else:
            new_row.append(0)
    z.append(list(new_row))

In [39]:
trace = go.Heatmap(z=z,
                   x=top_countries,
                   y=top_activities, xgap =5, ygap=5,
                  colorscale=[[0.0, 'rgb(199,204,118)'], 
                              [0.035,'rgb(69,117,180)' ],
                              [0.06, 'rgb(116,173,209)'],
                              [0.085, 'rgb(171,217,233)'],
                              [0.11, 'rgb(224,243,248)'],
                              [0.135, 'rgb(254,224,144)'],
                              [0.16, 'rgb(254,207,173)'],
                              [0.185, 'rgb(244,109,67)'],
                              [0.2, 'rgb(215,48,39)'],
                              [1.0, 'rgb(165,0,38)' ]])
data=[trace]


layout = go.Layout(
    title='Number of Shark attacks by Activity and Country',
    xaxis = dict(ticks='', nticks=20, title = 'Country', tickmode = 'linear'),
    yaxis = dict(ticks='', tickprefix = "", side='left',
        position=0.0, title = 'Activity', tickangle=0, tickfont=dict(
            size=10)))

fig = go.Figure(data=data, layout=layout)
fig['layout']['yaxis']['autorange'] = "reversed"
iplot(fig, filename = "Sharks heatmap")

It seems that surfing in the USA and Australia is by far the most dangerous activity. 
Sharks have been known to attack humans when they are confused or curious. Sharks often mistake surfers or a human splashing in the water for prey. They get curious and may try to investigate.
<br>
Here is the picture of how surfer on the board looks similar to shark’s prey:

In [40]:
Image(url= "plots/sealion_vs_surfer.png", width=500)

## Number of shark attacks by month and hemisphere

Let's find out ehich months the most shark attack happen. I am going to plot attacks by month, for that I have assigned to the southern or northen hemisphear since the distribution by hemisphere is different

In [48]:
months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

In [49]:
south = sharks[sharks.Hemisphere==1].groupby('Month').count().Date.values
north = sharks[sharks.Hemisphere==0].groupby('Month').count().Date.values

In [50]:
trace1 = go.Scatter(
    x=months,
    y=south,
    fill='tonexty',
    mode= 'none', name='Southern Hemisphere')
trace2 = go.Scatter(
    x=months,
    y=north,
    fill='tozeroy',
    mode= 'none', name='Northern Hemisphere')

data = [trace1, trace2]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Month'),
                   
    title='Number of shark attacks by month and hemisphere',
    yaxis=dict(
        title='Count of total shark attacks'),
    bargap=0.1)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

## Number of sharks by months, type and hemisphere

Another common misconception is that shark attacks increase when sharks are migrating or pupping. For example, shark pupping season along SoCal coast is known to be around April and May and shark migration from SoCal to Central and South America around winter time. 
<br>I wanted to know if it is true that the attacks are increasing due to migration and pupping, so I grouped attacks by months and separated countries into Northern and Southern Hemisphere as the distribution for these two groups would be different by months.

In [51]:
types = [i.lower() for i in sharks.Type.unique()]
types = [i.replace(' ', '_') for i in types]

for i, t in enumerate(types):
    vars()[types[i]+'_s'] = sharks[(sharks.Hemisphere==1) & (sharks.Type == sharks.Type.unique()[i]) & \
                                   (sharks.Year < 2018)]\
    .groupby('Month').count().Date.values

In [52]:
for i, t in enumerate(types):
    vars()[types[i]+'_n'] = sharks[(sharks.Hemisphere==0) & (sharks.Type == sharks.Type.unique()[i]) & \
                                   (sharks.Year < 2018)]\
    .groupby('Month').count().Date.values

In [53]:
types

['boat', 'unprovoked', 'invalid', 'provoked', 'questionable', 'sea_disaster']

In [54]:
trace0 = dict(
    x=months,
    y=invalid_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = pink),
    stackgroup='one', name = 'Invalid')
trace1 = dict(
    x=months,
    y=questionable_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = light_blue),
    stackgroup='one', name="Questionable")
trace2 = dict(
    x=months,
    y=sea_disaster_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = green),              
    stackgroup='one', name='Sea Disaster')
trace3 = dict(
    x=months,
    y=boat_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = dark_blue),              
    stackgroup='one', name='Boat')
trace4 = dict(
    x=months,
    y=provoked_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = red),           
    stackgroup='one', name='Provoked')
trace5 = dict(
    x=months,
    y=unprovoked_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = orange),             
    stackgroup='one', name='Unprovoked')
data = [trace0, trace1, trace2, trace3, trace4, trace5]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Month'),                  
    title='Number of shark attacks by months in Southern Hemisphere 1900-2017',
    yaxis=dict(
        title='Count of total shark attacks'),
    bargap=0.1)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

In [55]:
trace0 = dict(
    x=months,
    y=invalid_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = pink),
    stackgroup='one', name = 'Invalid')
trace1 = dict(
    x=months,
    y=questionable_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = light_blue),
    stackgroup='one', name="Questionable")
trace2 = dict(
    x=months,
    y=sea_disaster_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = green),              
    stackgroup='one', name='Sea Disaster')
trace3 = dict(
    x=months,
    y=boat_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),marker = dict(color = dark_blue),             
    stackgroup='one', name='Boat')
trace4 = dict(
    x=months,
    y=provoked_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = red),          
    stackgroup='one', name='Provoked')
trace5 = dict(
    x=months,
    y=unprovoked_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = orange),              
    stackgroup='one', name='Unprovoked')
data = [trace0, trace1, trace2, trace3, trace4, trace5]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Month'),                  
    title='Number of shark attacks by months in Northern Hemisphere 1900-2017',
    yaxis=dict(
        title='Count of total shark attacks'),
    bargap=0.1)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

Most attacks happen when people are on vacation and enjoying summer months in both Northern and Southern hemispheres and has very less to do with shark migration and pupping seasons. 

<i>Following graphs were generated in Tableau:</i>

## Bubble Map: Fatal shark attacks in 2008-2018

We know that the shark's population is decreasing, thus the patterns in the attacks might be different now than a couple of decades ago. Therefore, in a couple of following graphs, I focused on the shark attacks during the last decade. 
<br>This bubble map shows the number of fatal shark attacks. From this graph, we can clearly see that the attacks are very rare. For example, in the US there were only 9 fatal shark attacks during the last 10 years.

In [56]:
Image(url= "plots/bubble_map.png")

## Chloropleth map: Shark attacks in the US 2008-2018

To dig a bit deeper on the shark attacks in the US during 2008-2018, I plotted state location and the number of attacks on the choropleth map. Here all attacks are included: fatal, non-fatal, and unknown fatality. The number of attacks is pretty low. The attacks happen in places where people tend to enjoy the sea the most: Florida and California.

In [57]:
Image(url= "plots/chloropleth.png")

## Tree map: Number of shark attacks by species 2008-2018

Let’s take a look, which species of sharks have a higher likelihood to attack humans. Most of the attacks don’t have identified species; thus they are excluded from this analysis. 

In [58]:
Image(url= "plots/tree_map.png")

You can see that from global shark attacks during the last decade, the white shark is leading followed by a bull shark. Interestingly, that third place is when shark involvement is unconfirmed. This is mostly the cases where the people drown and sharks ate the cadaver or there was some other animal attacking such as pinniped.
<br>Just to give you a perspective, about 30-50 dies in the US yearly as a consequence of dog attack. Which is about 300-500 deaths in the US only for 2008-2018. This is more than double than the white shark attack worldwide, which includes non-fatal and unknown outcome.

## Connection map: Migration routes of sharks

Since my data doesn't really have any basis for a connection map, I decided to plot the major migration routes for sharks. These routes are based on my knowledge and internet research and do not represent any scientific point of view. There is a Pacific migration, happening from central California coast to other feeding grounds far away in the Pacific Ocean. Another major migration route is along the East coast, where sharks are migrating from the north to the warmer waters of Florida.

In [59]:
pacific_migration = [go.Scattergeo(
    lat = [28.822418, 38.170194],
    lon = [-158.859361, -123.720130],
    mode = 'lines',
    line = go.scattergeo.Line(
        width = 2,
        color = 'red',),)]

atlantic_migration = [go.Scattergeo(
    lat = [25.869109, 44.873876],
    lon = [-78.021723, -54.650979],
    mode = 'lines',
    line = go.scattergeo.Line(
        width = 2,
        color = 'red',),)]
layout = go.Layout(
    title = go.layout.Title(
        text = 'Approximate shark migration routes in Pacific and Atlantic side of the US'),
    showlegend = False,
    geo = go.layout.Geo(
        resolution = 50,
        showland = True,
        showlakes = True,
        landcolor = 'rgb(102, 153, 204)',
        countrycolor = 'rgb(102, 153, 204)',
        lakecolor = 'rgb(255, 255, 255)',
        projection = go.layout.geo.Projection(
            type = "equirectangular"),
        coastlinewidth = 2,
        lataxis = go.layout.geo.Lataxis(
            range = [20, 60],
            showgrid = True,
            dtick = 10),
        lonaxis = go.layout.geo.Lonaxis(
            range = [-100, 20],
            showgrid = True,
            dtick = 20
        ),
    )
)
fig = go.Figure(data = pacific_migration+atlantic_migration, layout = layout)
iplot(fig, filename = "Sharks stacked")

# How dangerous are sharks really?

While I had various interesting plots, I would like to focus on a stacked area graph, which supports my insight the most.  I am showing the shark attack count by month and fatality. 
<br> The fatality numbers are broken down by activity: surfing and other activity. To account for the difference in seasons, I grouped countries into Northern and Southern hemispheres. For the time period, I decided to explore a period of 10 years. But first, <br> I would like to start with supporting chart: bar plot by hemisphere, which shows that there are far more shark attacks in Northern hemisphere than in Southern, although deadly attacks are twice as likely in the Southern hemisphere than in the Northern.


In [60]:
sharks['Fatal (Y/N)'].unique()

array(['N', 'Y', 'UNKNOWN'], dtype=object)

In [61]:
south_u = sharks[(sharks.Hemisphere==1) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='UNKNOWN')].count()[0]
south_n = sharks[(sharks.Hemisphere==1) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='N')].count()[0]
south_y = sharks[(sharks.Hemisphere==1) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='Y')].count()[0]
north_u = sharks[(sharks.Hemisphere==0) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='UNKNOWN')].count()[0]
north_n = sharks[(sharks.Hemisphere==0) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='N')].count()[0]
north_y = sharks[(sharks.Hemisphere==0) & (sharks.Year.isin(list(range(2007, 2017, 1)))) 
               & (sharks['Fatal (Y/N)']=='Y')].count()[0]

In [62]:
percentage_nonfatal = (south_n+north_n)/(south_n +south_u+ south_y+north_n+north_u+north_y)*100
f'Percentage of Non-Fatal shark attacks {round(percentage_nonfatal, 2)}%'

'Percentage of Non-Fatal shark attacks 84.36%'

In [63]:
trace1 = go.Bar(
    x=['Southern', 'Northern'],
    y=[south_u, north_u], marker = dict(color=light_blue),
    name='Unknown')
trace2 = go.Bar(
    x=['Southern', 'Northern'],
    y=[south_y, north_y],
    name='Fatal', marker = dict(color=pink))
trace3 = go.Bar(
    x=['Southern', 'Northern'],
    y=[south_n, north_n],
    name='Non-Fatal', marker = dict(color=green))
data = [trace1, trace2, trace3]
layout = go.Layout(
    barmode='stack', title = 'Shark attacks by hepisphere 2007-2017', xaxis=dict(title='Hemisphere'),
    yaxis=dict(
        title='Count of total shark attacks'),)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

From the two stacked plots below, you can see that generally, fatal shark attacks are very rare, especially in the Northern Hemisphere, although the overall there are more attacks in the Northern Hemisphere.
<br>The fatality rate is higher in the Southern Hemisphere, the number of attacks is higher in total for Northern Hemisphere.  Nearly half of the attacks happen while people are surfing, the situation where the shark mistake surfers for their favorite prey - seals.

In [67]:
fatality = [i.lower() for i in sharks['Fatal (Y/N)'].unique()]

In [68]:
for i, t in enumerate(fatality):
    vars()[fatality[i]+'_s'] = sharks[(sharks.Hemisphere==1) & (sharks['Fatal (Y/N)'] == sharks['Fatal (Y/N)'].unique()[i]) & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values

In [69]:
fatal_surfing = sharks[(sharks.Hemisphere==1) & (sharks['Fatal (Y/N)'] == 'Y') & (sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))].groupby('Month').count().Date.values

fatal_not_surfing = sharks[(sharks.Hemisphere==1) & (sharks['Fatal (Y/N)'] == 'Y') &(sharks.Activity != 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values
    
nonfatal_surfing = sharks[(sharks.Hemisphere==1) & (sharks['Fatal (Y/N)'] == 'N') &(sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values
nonfatal_not_surfing = sharks[(sharks.Hemisphere==1) & (sharks['Fatal (Y/N)'] == 'N') &(sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values

trace0 = dict(
    x=months,
    y=unknown_s,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = dark_blue),
              
    stackgroup='one', name = 'Unknown'
)
trace1 = dict(
    x=months,
    y=nonfatal_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = light_blue),
    stackgroup='one', name="Non-Fatal Surfing"
)
trace2 = dict(
    x=months,
    y=nonfatal_not_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = green),
              
    stackgroup='one', name='Non-Fatal Other'
)

trace3 = dict(
    x=months,
    y=fatal_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = red),
    stackgroup='one', name="Fatal Surfing"
)
trace4 = dict(
    x=months,
    y=fatal_not_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = orange),
              
    stackgroup='one', name='Fatal Other'
)

data = [trace0, trace1, trace2, trace3, trace4]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Month'),
                   
    title='Number of shark attacks by months in Southern Hemisphere 2007-2017',
    yaxis=dict(
        title='Count of total shark attacks'
    ),
    bargap=0.1
)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

In [70]:
for i, t in enumerate(fatality):
    vars()[fatality[i]+'_n'] = sharks[(sharks.Hemisphere==0) & (sharks['Fatal (Y/N)'] == sharks['Fatal (Y/N)'].unique()[i]) & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values
    
fatal_surfing = sharks[(sharks.Hemisphere==0) & (sharks['Fatal (Y/N)'] == 'Y') & (sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))].groupby('Month').count().Date.values

fatal_not_surfing = sharks[(sharks.Hemisphere==0) & (sharks['Fatal (Y/N)'] == 'Y') &(sharks.Activity != 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values
    
nonfatal_surfing = sharks[(sharks.Hemisphere==0) & (sharks['Fatal (Y/N)'] == 'N') &(sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values
nonfatal_not_surfing = sharks[(sharks.Hemisphere==0) & (sharks['Fatal (Y/N)'] == 'N') &(sharks.Activity == 'Surfing') & \
                                   (sharks.Year.isin(list(range(2007, 2017, 1))))]\
    .groupby('Month').count().Date.values

trace0 = dict(
    x=months,
    y=unknown_n,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5), marker = dict(color = dark_blue),
              
    stackgroup='one', name = 'Unknown'
)
trace1 = dict(
    x=months,
    y=nonfatal_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = light_blue),
    stackgroup='one', name="Non-Fatal Surfing"
)
trace2 = dict(
    x=months,
    y=nonfatal_not_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = green),
              
    stackgroup='one', name='Non-Fatal Other'
)

trace3 = dict(
    x=months,
    y=fatal_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = red),
    stackgroup='one', name="Fatal Surfing"
)
trace4 = dict(
    x=months,
    y=fatal_not_surfing,
    hoverinfo='x+y',
    mode='lines',
    line=dict(width=0.5),  marker = dict(color = orange),
              
    stackgroup='one', name='Fatal Other'
)

data = [trace0, trace1, trace2, trace3, trace4]

layout = go.Layout(barmode='overlay',
        xaxis=dict(title='Month'),
                   
    title='Number of shark attacks by months in Northern Hemisphere 2007-2017',
    yaxis=dict(
        title='Count of total shark attacks'
    ),
    bargap=0.1
)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

# Summary

While a shark attack is a frightening event, we need to understand that it is very rare and happens by mistake. Most of the shark accidents occur when the shark feels threatened or confused by either the presence of many dead fishes, murky water or mistakes people for food. 
<br>Surfers, swimmers, and fishers in the USA and Australia are the most common case for the shark attacks. Though in 84.4 % the attack is not fatal.
<br>Many of the attacks are very minor cases recorded such as when shark bumps the boat. There are also many cases where shark attack is not confirmed, either the attack is by other animal or the shark attack happened post-mortem in case of drowning.
<br>I hope this data and my visualizations could convince you that sharks are not the killer machines waiting for you to enter the waters, but just curious animals which need to be treated with respect.

**Sources**
<br>https://www.projectaware.org/sharks
<br>https://www.sharkwater.com/
<br>https://en.m.wikipedia.org/wiki/Fatal_dog_attacks_in_the_United_States
<br>https://www.dailymail.co.uk/news/Researchers-examine-sharks-mistake-surfers-intended-prey

# Bonus: California shark attacks 2007 - 2017

As a scuba diver in California, I wanted to know what is the statistics of shark attacks in the state for the last 10 years

In [79]:
num_attacks = sharks[(sharks.Area == 'California')& sharks.Year.isin(list(range(2007, 2018, 1)))].count()[0]
f'There are total {num_attacks} recorded shark attacks in California in the last 10 yeas'

'There are total 86 recorded shark attacks in California in the last 10 yeas'

Let's break it down by fatality, activities and type

In [150]:
sharks_ca = sharks[(sharks.Area == 'California')& sharks.Year.isin(list(range(2007, 2018, 1)))]
sharks_ca.head()

Date  Year        Type Country        Area  \
54   30-Dec-2017  2017  Unprovoked     USA  California   
60   24-Nov-2017  2017  Unprovoked     USA  California   
105  01-Aug-2017  2017        Boat     USA  California   
116  20-Jul-2017  2017        Boat     USA  California   
117  20-Jul-2017  2017  Unprovoked     USA  California   

                                              Location      Activity Sex  Age  \
54            Drakes Estero, Point Reyes, Marin County       Surfing   F   35   
60                    Stillwater Cove, Monterey County  Spearfishing   M   25   
105  Between Pescadero Point & Bean Hollow Beach, S...       Fishing   M  NaN   
116                       Stearns Wharf, Santa Barbara      Kayaking   M   40   
117             Seal Rock, Goleta Beach, Santa Barbara           SUP   M   50   

                               Injury Fatal (Y/N)   Time      Species Hour  \
54                        Foot bitten           N  12h00          NaN   12   
60                   Right leg bitten           N  13h45  White shark   13   
105    No injury, bow of kayak bitten           N  10h45  White shark   10   
116           No injury, kayak bitten           N  11h20  White shark   11   
117  No injury, shark bit paddleboard           N  07h45  White shark    7   

    Month Hemisphere  
54     11          0  
60     10          0  
105     7          0  
116     6          0  
117     6          0

In [142]:
sharks_ca.Activity.unique()

array(['Surfing', 'Spearfishing', 'Fishing', 'Kayaking', 'SUP',
       'Swimming', 'Kayak fishing', 'Treading water', nan, 'Canoeing',
       'Paddle boarding', 'Body boarding', 'Surf paddling'], dtype=object)

In [156]:
top_activities_ca = list(sharks_ca.groupby('Activity').count().sort_values('Date', ascending=False)[:20].index)

In [157]:
heatmap_ca = sharks_ca[sharks_ca.Activity.isin(top_activities_ca)]
heatmap_ca['Count'] = 1
heatmap_ca = heatmap_ca[['Month', 'Activity', 'Count']]
a = heatmap_ca.groupby(['Month', 'Activity'], group_keys=False).sum()

heatmap_ca_dict = {}

for i in a.itertuples():
    heatmap_ca_dict[i[0]] =i[1]

In [158]:
month_dict = {}
for index, m in enumerate(months):
    month_dict[index] = m

In [159]:
z_ca = []

for act in top_activities_ca:
    new_row = []
    for m in month_dict.keys():
        if ((m), str(act)) in heatmap_ca_dict:
            new_row.append(heatmap_ca_dict[((m), str(act))])
        else:
            new_row.append(0)
    z_ca.append(list(new_row))

In [207]:
trace = go.Heatmap(z=z_ca,
                   x=months,
                   y=top_activities_ca, xgap =5, ygap=5,
                  colorscale=[[0.0, 'rgb(199,204,118)'], 
                              [0.035,'rgb(69,117,180)' ],
                              [0.06, 'rgb(116,173,209)'],
                              [0.085, 'rgb(171,217,233)'],
                              [0.11, 'rgb(224,243,248)'],
                              [0.135, 'rgb(254,224,144)'],
                              [0.16, 'rgb(254,207,173)'],
                              [0.185, 'rgb(244,109,67)'],
                              [0.2, 'rgb(215,48,39)'],
                              [1.0, 'rgb(165,0,38)' ]])
data=[trace]


layout = go.Layout(
    title='Shark Attacks by Activity and Month in California 2007-2018',
    xaxis = dict(ticks='', nticks=20, title = 'Month', tickmode = 'linear'),
    yaxis = dict(ticks='', tickprefix = "", side='left',
        position=0.0, tickangle=0, tickfont=dict(
            size=10)))

fig = go.Figure(data=data, layout=layout)
fig['layout']['yaxis']['autorange'] = "reversed"
iplot(fig, filename = "Sharks heatmap")

There are a total of 86 attacks. There were no attacks on scuba divers in California for the last 10 years. Most attacks are on Surfers and Fishers.

In [188]:
sharks_ca['Fatal (Y/N)'].unique()

array(['N', 'UNKNOWN', 'Y'], dtype=object)

Let's inverstigate the fatality and injury

In [189]:
fatal = []
non_fatal = []
unknown = []
for f in sharks_ca['Fatal (Y/N)'].unique():

    vars()[f] = []
    for act in top_activities_ca[:6]:

        vars()[f].append(sharks_ca[(sharks_ca['Fatal (Y/N)']==f) & (sharks_ca.Activity==act)].count()[0])

In [190]:
UNKNOWN

[6, 0, 0, 2, 0, 0]

In [191]:
trace1 = go.Bar(
    x=top_activities_ca[:6],
    y=UNKNOWN, marker = dict(color=light_blue),
    name='Unknown')
trace2 = go.Bar(
    x=top_activities_ca[:6],
    y=Y,
    name='Fatal', marker = dict(color=pink))
trace3 = go.Bar(
    x=top_activities_ca[:6],
    y=N,
    name='Non-Fatal', marker = dict(color=green))

data = [trace1, trace2, trace3]
layout = go.Layout(
    barmode='stack', title = 'Shark attacks in California 2007-2017 by Activity', xaxis=dict(title='Activity'),
    yaxis=dict(
        title='Count of shark attacks'),)

fig = go.Figure(data=data, layout=layout)
iplot(fig, filename = "Sharks stacked")

You can see that there is only 3 fatal case in California for the last 10 years.

In [195]:
sharks_ca[sharks_ca['Fatal (Y/N)']=='Y'].count()[0]

3

Let's get rid of all no injury attacks to calculate the percentage of accidents with a shark which led to injury

In [197]:
invol = []
col = 'Injury'
word = 'No injury'
for i in sharks_ca.Injury.iteritems():
    try:
        x = re.findall(f'({word})|({word.lower()})', i[1])
        if len(x)>0:
            invol.append(i[1])
    except:
        pass
set(invol)

{' No injury. Shark bumped kayak, flinging her into the  water. ',
 'Board reportedly bumped by shark. No injury',
 'No injury',
 'No injury to occupant, shark/s holded  kayak',
 'No injury to occupant. Kayak bitten by gaffed shark. PROVOKED INCIDENT',
 'No injury to surfer, board bitten',
 'No injury to surfer, surfboard bitten',
 'No injury to surfer, surfboard bitten by the shark',
 'No injury, board bitten',
 'No injury, board broken in half',
 'No injury, board damaged',
 'No injury, bow of kayak bitten',
 'No injury, kayak bitten',
 'No injury, kayak damaged',
 'No injury, kayak scratched',
 'No injury, kayaker fell in the water when kayak bitten by a shark',
 'No injury, minor damage to wetsuit',
 'No injury, shark bit paddleboard',
 'No injury, shark bit surfboard',
 'No injury, shark bumped & damaged board',
 'No injury, shark bumped surfboard',
 'No injury, shark collided with surfer & board',
 'No injury, shark rammed kayak repeatedly',
 'No injury, shark struck paddle',
 "N

In [200]:
no_inj = len(invol)
no_inj

44

In [204]:
1-no_inj/num_attacks

0.4883720930232558

In [205]:
3/num_attacks

0.03488372093023256

Many attacks have no injury at all, this is where shark only bumps a kayak or a surfboard. If I exclude those ‘no injury’ attacks we can see that only 48.8% of all 86 shark attacks in CA in the past 10 years led to injury and only 3.4% of them were deadly.